option to include colors of Bloch vectors and points

rajathshetty20 opened the pull request:
**Description**
Added option to include colors to Bloch vectors and points instead of just cycling from a list 

**Related issues or PRs**
fix #1292 

**Changelog**
Added an optional argument called colors for the Bloch.add_vectors, Bloch.add_points and Bloch.add_states method. 
Tested it here: https://github.com/rajathshetty20/misc/blob/master/qutip%20%231292.ipynb

nathanshammah commented:
maybe @quantshah or @araza6 can have a look at this?

sarsid commented:
Just a small suggestion, I am not too familiar with the details here but maybe this can also be used/demonstrated in the notebook [Bloch Sphere with Colorbar](https://nbviewer.jupyter.org/github/qutip/qutip-notebooks/blob/master/examples/bloch_sphere_with_colorbar.ipynb)

rajathshetty20 commented:
Can someone review this? I'm new to open-source. Is my implementation right?

rajathshetty20 commented on qutip/bloch.py:
```diff
@@ -375,19 +377,27 @@ def add_states(self, state, kind='vector'):
         if isinstance(vectors[0], (list, ndarray)):
-            for vec in vectors:
-                self.vectors.append(vec)
+            for k in range(len(vectors)):
```
Okay, i will add the color option to add points also, and use enumerate. Could you suggest a way of doing this without using a dictionary? As the method where the vector is created and the method where the vectors are plotted is different, i will have to store the vector index and color somewhere. That is why I used a dictionary.

rajathshetty20 commented:
@Ericgig I have added the color option for add_points also. For the single-color method, a single color with be passed. And for the multi-color method, a list of colors will be passed. I have tested it in https://github.com/rajathshetty20/misc/blob/master/qutip%20%231292.ipynb

Ericgig commented:
You can start with `self.vector_color = []` in `__init__` and append a color each time you append a vector. If there is no color given, the next one from the default list is added in `add_vectors` and `add_points`. By keeping `self.vector_color` and `self.vectors` the same length, dictionary are not needed.
But lets keep the `mod(k, len(self.vector_color))` in case `self.vector_color` is overwritten by the some users. 

`add_states` calls `add_vectors` and `add_points`, it could have a color argument to pass to them.

rajathshetty20 commented:
@Ericgig I removed the dictionary, and added lists called vector_display_color and points_display_color. It works the same as in the previous version (https://github.com/rajathshetty20/misc/blob/master/qutip%20%231292.ipynb)

Ericgig commented:
@rajathshetty20, 
One last thing, please rename
`self.vector_color` to `self.vector_default_color`
`self.vector_display_color`  to `self.vector_color`
and the same for points.
So the [example](https://nbviewer.jupyter.org/github/qutip/qutip-notebooks/blob/master/examples/bloch_sphere_with_colorbar.ipynb) displays the color as expected.

rajathshetty20 commented:
@Ericgig I changed the variable names. In the example file one small change has to be made. For changing the b.point_color after points are added, we will have to set b.point_color = [colors] instead of b.point_colors = list(colors). b.point_color is a list of list of points.  https://github.com/rajathshetty20/misc/blob/master/bloch_sphere_with_colorbar.ipynb

Ericgig commented:
I would prefer it not breaking any code with master. Ideally, we should raise a deprecation warning first, then upload the change later... There is a major new version in development that would allow this kind of changes in the branch `dev.major`. 
So could you move this PR to merge to `dev.major` instead of `master`. 

If you prefer not to wait for the next major release, which will take some time, you can remove the list of list in `Bloch.point_colors` so there is no need to change the examples.

rajathshetty20 commented:
@Ericgig I could store point_colors as a list of colors for all the points. But storing it as list of lists is more consistent with the way points, and point styles are stored. So, i think it is better to keep it this way. I am changing the target branch to `dev.major`

rajathshetty20 commented:
@Ericgig Now, that I have changed the 'merge to' branch. This PR contains a lot of the other commits in master branch which are not yet merged with the dev,major branch. so should i create a new PR starting from the dev.major branch

Ericgig commented:
Yes, it would be easier that way, master and dev.major split a while ago. The bloch.py file have not been touched at all so it should be simple.

rajathshetty20 commented:
@Ericgig I created a new PR to merge with dev.major #1335
